File: simplicity/src/simplicity/engine/RunnableEngine.cpp
/*
 This file is part of The Simplicity Engine.

 The Simplicity Engine is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published
 by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

 The Simplicity Engine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with The Simplicity Engine. If not, see <http://www.gnu.org/licenses/>.
 */
#include <exception>
#include <string>
using namespace std;

#include <boost/date_time.hpp>
#include <boost/thread.hpp>
#include <log4cpp/Category.hh>
using namespace boost;
using namespace boost::posix_time;
using namespace boost::this_thread;
using namespace log4cpp;

#include "../SENotSupportedException.h"
#include "RunnableEngine.h"

namespace simplicity
{
  Category * RunnableEngine::fLogger = &Category::getInstance("simplicity::RunnableEngine");

  RunnableEngine::RunnableEngine() :
    fPreferredFrequency(1), fSleepTime(0L)
  {
  }

  RunnableEngine::~RunnableEngine()
  {
  }

  int
  RunnableEngine::getPreferredFrequency() const
  {
    return (fPreferredFrequency);
  }

  void
  RunnableEngine::init()
  {
    initInternal();
  }

  void
  RunnableEngine::initInternal()
  {
    fInterrupted = false;
    fSleepTime = (long) (MILLISECONDS_IN_A_SECOND / fPreferredFrequency);
  }

  void
  RunnableEngine::run()
  {
    init();

    // Start by sleeping.
    ptime beforeAdvanceTime;
    long adjustedSleepTime = sleep(fSleepTime);

    // While the engine has not been interrupted.
    while (!fInterrupted)
      {
        // Interrupt the engine if the thread has been interrupted.
        interruption_point();

        beforeAdvanceTime = microsec_clock::local_time();

        try
          {
            advance(NULL);
          }
        catch (std::exception& e)
          {
            // Interrupt the engine.
            fInterrupted = true;
            fLogger->error("Failed to advance the engine.");
          }

        // Subtract the time taken to advance the engine from the time it needs to sleep for.
        adjustedSleepTime -= time_period(beforeAdvanceTime, microsec_clock::local_time()).length().total_milliseconds();

        // Sleep until the next advancement is due.
        adjustedSleepTime = sleep(adjustedSleepTime);
      }

    destroy();
  }

  void
  RunnableEngine::setPreferredFrequency(int const preferredFrequency)
  {
    fPreferredFrequency = preferredFrequency;
  }

  long
  RunnableEngine::sleep(long const adjustedSleepTime)
  {
    // If the engine needs to sleep.
    if (adjustedSleepTime > 0)
      {
        try
          {
            this_thread::sleep(milliseconds(adjustedSleepTime));
          }
        catch (thread_interrupted& e)
          {
            // Interrupt the engine.
            fInterrupted = true;
            fLogger->debug("The engine was interrupted while sleeping.");
          }

        // Return the standard sleep duration.
        return (fSleepTime);
      }
    else
      {
        fLogger->warn("The engine ran over time.");

        // The engine has not slept as a result it has 'caught up' by the standard sleep duration.
        return (adjustedSleepTime + fSleepTime);
      }
  }
}

File: com.se.simplicity/src/com/se/simplicity/engine/RunnableEngine.java
package com.se.simplicity.engine;

import org.apache.log4j.Logger;

/**
 * <p>
 * An engine that can be run at the preferred frequency provided to it. If an advancement takes longer than the time allowed by the preferred
 * frequency, this <code>RunnableEngine</code> will attempt to 'catch up' by advancing continuously until the correct number of advancements have been
 * made for the time this <code>RunnableEngine</code> has been running.
 * </p>
 * 
 * <p>
 * <code>RunnableEngine</code> is designed to be run in a new thread. It implements {@link java.lang.Runnable Runnable} and is cleanly interruptible.
 * For more information on Java threading see the tutorials on the Java web site.
 * </p>
 * 
 * @author Gary Buyn
 */
public abstract class RunnableEngine implements Engine, Runnable
{
    /**
     * <p>
     * Logs messages associated with this class.
     * </p>
     */
    private static Logger fLogger = Logger.getLogger(RunnableEngine.class);

    /**
     * <p>
     * The preferred frequency (advancements per second) of this <code>RunnableEngine</code>.
     * </p>
     */
    private int fPreferredFrequency;

    /**
     * <p>
     * The unadjusted time for this <code>RunnableEngine</code> to sleep between advancements. This <code>RunnableEngine</code> would only sleep this
     * long if the previous advancement was instantaneous.
     * </p>
     */
    private long fSleepTime;

    /**
     * <p>
     * Creates an instance of <code>RunnableEngine</code>.
     * </p>
     */
    public RunnableEngine()
    {
        fPreferredFrequency = 1;
        fSleepTime = 0L;
    }

    @Override
    public void destroy()
    {}

    @Override
    public int getPreferredFrequency()
    {
        return (fPreferredFrequency);
    }

    @Override
    public void init()
    {
        fSleepTime = (long) (MILLISECONDS_IN_A_SECOND / fPreferredFrequency);
    }

    @Override
    public void run()
    {
        init();

        // Start by sleeping.
        long beforeAdvanceTime = 0;
        long adjustedSleepTime = sleep(fSleepTime);

        // While the engine has not been interrupted.
        while (!Thread.interrupted())
        {
            beforeAdvanceTime = System.currentTimeMillis();

            try
            {
                advance(null);
            }
            catch (Exception e)
            {
             // Interrupt the engine.
                Thread.currentThread().interrupt();
                fLogger.error("Failed to advance the engine.", e);
            }

            // Subtract the time taken to advance the engine from the time it needs to sleep for.
            adjustedSleepTime -= System.currentTimeMillis() - beforeAdvanceTime;

            // Sleep until the next advancement is due.
            adjustedSleepTime = sleep(adjustedSleepTime);
        }

        destroy();
    }

    @Override
    public void setPreferredFrequency(final int preferredFrequency)
    {
        fPreferredFrequency = preferredFrequency;
    }

    /**
     * <p>
     * Causes this <code>RunnableEngine</code> to sleep for the adjusted time given (assuming it is it's own thread). The adjusted sleep time is then
     * updated to account due to the fact that this <code>RunnableEngine</code> has just slept.
     * </p>
     * 
     * <p>
     * In the case that the adjusted sleep time was positive, it is simply reset to the regular sleep time. In the case that it was negative (or
     * zero), the regular sleep time is added to it (and no delay is actually effected).
     * </p>
     * 
     * @param adjustedSleepTime The adjusted time this <code>RunnableEngine</code> is required to sleep. The 'adjusted time' is the time taken to
     * execute the previous advancement subtracted from the 'regular time' (the time between advancements).
     * 
     * @return The updated adjusted sleep time.
     */
    private long sleep(final long adjustedSleepTime)
    {
        // If the engine needs to sleep.
        if (adjustedSleepTime > 0)
        {
            try
            {
                Thread.sleep(adjustedSleepTime);
            }
            catch (InterruptedException e)
            {
                // Interrupt the engine.
                Thread.currentThread().interrupt();
                fLogger.debug("The engine was interrupted while sleeping.");
            }

            // Return the standard sleep duration.
            return (fSleepTime);
        }
        else
        {
            fLogger.warn("The engine ran over time.");

            // The engine has not slept as a result it has 'caught up' by the standard sleep duration.
            return (adjustedSleepTime + fSleepTime);
        }
    }
}
